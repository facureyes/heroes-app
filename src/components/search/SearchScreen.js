import React, { useMemo, useState } from 'react'
import queryString from 'query-string'
import { useLocation } from 'react-router-dom'
import HeroeCard from '../heroes/HeroeCard'
import { getHeroesByName } from '../../selectors/getHeroesByName'

const SearchScreen = ({history}) => {
  const location = useLocation();
  const { q = ''} =queryString.parse(location.search);
  const [search, setSearch] = useState(q);
  const heroesFiltered = useMemo(() => getHeroesByName(q), [q])
  


  const handleSubmit = event => {
    event.preventDefault(); 
    if(search !== ''){
      console.log(heroesFiltered)
      history.push(`?q=${search}`);
    } else {
      history.push('/search');
    }
  }
  
  return (
    <div>
      <h1>Search Screen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input 
              type="text"
              name="searchText"
              autoComplete="off"
              placeholder="Find your hero"
              className="form-control"
              onChange={(e)=>setSearch(e.target.value)}
              value={search}
            />
            <button
              type="submit"
              className="btn mt-2 btn-block btn-outline-dark"
            >
              Search
            </button>
          </form>

        </div>
        <div className="col-7">
          <h4>Results</h4>
          <hr />

          {q==='' && 
          <div className="alert alert-info">
            Search a Hero.
          </div>}
          
          {(heroesFiltered.length < 1 && q !== '') && 
          <div className="alert alert-danger">
            There is no hero with "{q}".
          </div>}

          {
            heroesFiltered.map(heroe => 
              <HeroeCard 
                key={heroe.id} 
                {...heroe} 
              />
            )
          }
        </div>

      </div>
    </div>
  )
}

export default SearchScreen
