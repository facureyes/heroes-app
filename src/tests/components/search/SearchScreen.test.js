import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router-dom";
import SearchScreen from "../../../components/search/SearchScreen"

describe('Tests on <SearchScreen />', () => {
  
  test('should show with default values', () => {
    const wrapper = mount(
    <MemoryRouter initialEntries={['/search']}>
      <Route path="/search" component={SearchScreen} />
    </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a Hero.');
  });
  
  test('should show Batman and input with queryString value', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=batman']}>
      <Route path="/search" component={SearchScreen} />
    </MemoryRouter>
    );

    expect(wrapper.find('input').prop('value')).toBe('batman');
    expect(wrapper.find('HeroeCard').prop('superhero').toLowerCase()).toBe('batman');
    expect(wrapper).toMatchSnapshot();
  });


  test('should show error if Hero isn\'t found', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search?q=salchipapa']}>
      <Route path="/search" component={SearchScreen} />
    </MemoryRouter>
    );

    expect(wrapper.find('.alert-danger').text().trim()).toBe('There is no hero with \"salchipapa\".');
  })
  

  test('should call push from history', () => {
    const history = {
      push: jest.fn(),
    };

    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
      <Route path="/search" component={() => <SearchScreen history={history} />} />
    </MemoryRouter>
    );

    wrapper.find('input').simulate('change', {
      target: {
        name: 'searchText',
        value: 'batman',
      }
    });

    wrapper.find('form').simulate('submit');

    expect(history.push).toHaveBeenCalledWith('?q=batman');
    

  })
  
})
