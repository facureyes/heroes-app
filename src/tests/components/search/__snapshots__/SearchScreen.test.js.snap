// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchScreen /> should show Batman and input with queryString value 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn mt-2 btn-block btn-outline-dark"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card mt-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/heroe/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests on <SearchScreen /> should show with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn mt-2 btn-block btn-outline-dark"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a Hero.
      </div>
    </div>
  </div>
</div>
`;
